### PR TITLE
Overview Documentation

### PR DESCRIPTION
**Pull request by colin-higgins**
Flowchart detailing how tests execute locally and in CI.
Architectural overview document.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,13 +10,30 @@ Workbench designed to run advanced tests (integration, smoke, functionnal, fuzzi
 
 Add a valid staging `DD_API_KEY` environment variable (you can set it in a `.env` file). Then:
 
-```bash
-./build.sh
-./run.sh
+```mermaid
+flowchart TD
+    BUILDNODE[./build.sh node] --> BUILT
```
Comment (cbeauchesne) [inline README.md:15]:
```suggestion
    BUILDNODE[./build.sh nodejs] --> BUILT
```

File: docs/architecture/overview.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    BACKEND[Datadog]
+```
+
+The tests send requests directly to the application.
+The tests then wait on the results, which are available as the logs are collected from [mitmproxy](TODO) dumps.
+
+## What are system-tests bad for?
+
+ - Combinatorial style tests
+ - Tests that require specific versions of runtimes, libraries, or operating systems
```
Comment (cbeauchesne) [inline docs/architecture/overview.md:47]: Even if it's not as easy as adding a test case of an existing test, it's kind of doable as long as it runs on docker. For instance, ruby engineer added lot of rails/sinatra version with a reasonable effort. 

I would say that even if it's not build to crawl a version matrix, it's made to be easy enough to add new frameworks/runtime, or relevants version of them as long as it runs on docker.

File: docs/architecture/overview.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+ - Add the enabling environment variable to the shared docker-compose
+ - Add a test asserting the environment variable affected traces as expected
+ - Build and run for each language to verify the tests run and behave as expected
+ - Create the pull request in system-tests
+
+However, there are many scenarios where a test may not be so simple to implement.
+
+This document aims to give a working understanding of the parts of system-tests, and how to troubleshoot them.
+
+## What are the components of a running test?
```
Comment (omerli) [inline docs/architecture/overview.md:17]: not sure if that's the right section, but it's probably also worth mentioning how tests can be integrated to the CI/CD process and where to find the results (link to the dashboard)

File: docs/architecture/overview.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+ - Combinatorial style tests
+ - Tests that require specific versions of runtimes, libraries, or operating systems
+ - Cloud deployments, kubernetes, distributed deployments
+ - Immediately knowing the reason a feature fails
+ - Problems or features which are not shared across tracers
+ - Performance or throughput testing
+
+ *Examples of bad candidates:*
+  - The .NET tracer must not write invalid IL for it's earliest supported runtime
```
Comment (omerli) [inline docs/architecture/overview.md:57]: nit: clarify IL acronym

File: docs/architecture/overview.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+## How do I add a new test?
+
+The default folder to add new tests is `./tests`.
+
+The framework used for running tests is [pytest](https://docs.pytest.org/).
+
+For a test to be run, it must have the filename prefix `test_`.
+
+Follow the example provided within `./tests/template_test.py` if you are unfamiliar with pytest.
```
Comment (omerli) [inline docs/architecture/overview.md:82]: i like the idea of a template.
nit: maybe add direct link to `./tests/template_test.py`

**Review by omerli — APPROVED** [review]

File: docs/architecture/overview.md (hunk 5 above)
Comment (colin-higgins) [reply docs/architecture/overview.md]: done

File: docs/architecture/overview.md (hunk 4 above)
Comment (colin-higgins) [reply docs/architecture/overview.md]: Done

File: docs/architecture/overview.md (hunk 3 above)
Comment (colin-higgins) [reply docs/architecture/overview.md:17]: I will do so.

File: docs/architecture/overview.md (hunk 2 above)
Comment (colin-higgins) [reply docs/architecture/overview.md]: Clarified this a bit

File: docs/architecture/overview.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+The build script must be successful before running the tests.
+The first argument to the `./run.sh` script is the scenario (`$SCENARIO`) which defaults to `DEFAULT`.
+    - `./run.sh`
+    - `./run.sh DEFAULT`
+    - `./run.sh SAMPLING`
+    - `./run.sh PROFLING`
+
+You can see all available scenarios within the `./run.sh` script.
+The run script sets necessary variables for each scenario, which are then used within the `docker-compose.yml` file.
+
```
[the remark this answers is not in the document]
Comment (colin-higgins) [reply docs/architecture/overview.md:158]: Added in another section

File: docs/architecture/overview.md (hunk 3 above)
Comment (colin-higgins) [reply docs/architecture/overview.md:17]: I added a section about how to test specific tracer instances but would like to do a follow up if that's okay.

File: docs/architecture/overview.md [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+The `./run.sh` script runs the tests and relies 1-to-1 on what is built in the `./build.sh` step.
+ - [Click for details about the `./run.sh` script and options available](#running-the-system-tests).
+
+The run script ultimately calls the `./docker-compose.yml` file and whichever image is built with the `weblog` tag is tested. 
+ - [Click for detail about how the images interact with eachother](#what-are-the-components-of-a-running-test)
+
+## Building the System Tests
+
+The first argument to the `./build.sh` script is the language (`$TEST_LIBRARY`) which is built: `./utils/build/docker/{language}`.
+ - `./build.sh ruby`
```
Comment (cbeauchesne) [inline docs/architecture/overview.md:131]:
```suggestion
 - `./build.sh cpp`
 - `./build.sh ruby`
```

File: docs/architecture/overview.md [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+ - `./build.sh python`
+ - `./build.sh php`
+ - `./build.sh nodejs`
+ - `./build.sh java`
+ - `./build.sh golang`
+ - `./build.sh dotnet`
+
+There are explicit arguments available for more specific configuration of the build.
+ - i.e., `./build.sh --library {language} --weblog-variant {dockerfile-prefix}`
+ - e.g., `./build.sh --library python --weblog-variant flask-poc`
```
Comment (cbeauchesne) [inline docs/architecture/overview.md:141]:
```suggestion
 - e.g., `./build.sh python --weblog-variant flask-poc`
 - shorter version: ./build.sh python -w flask-poc

```

File: tests/template_test.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
+# Unless explicitly stated otherwise all files in this repository are licensed under the the Apache License Version 2.0.
+# This product includes software developed at Datadog (https://www.datadoghq.com/).
+# Copyright 2021 Datadog, Inc.
+
+from utils import BaseTestCase, interfaces, context, irrelevant
+from utils.warmups import default_warmup
+
+# This warmup task runs to make sure containers are ready to receive requests
+context.add_warmup(default_warmup)
```
Comment (cbeauchesne) [inline tests/template_test.py:9]: You can remove this line, it's only for dedicated scenarios

Comment (colin-higgins) [reply tests/template_test.py:9]: Maybe it's good to leave it in but comment it out? and explain what it is for?

File: docs/architecture/overview.md [hunk 10]
```diff
@@ -0,0 +1,201 @@
+# Overview
+
+The components that make up a running test are simple from the outside.
+
+The idea behind system tests is that we can share the tests for a given feature across implementations.
+
+Enabling a feature within system tests might go like this:
```
[the remark this answers is not in the document]
Comment (colin-higgins) [reply docs/architecture/overview.md:7]: The idea here was to try to run for each language, but I suppose CI can do this.

[the remark this answers is not in the document]
Comment (colin-higgins) [reply docs/architecture/overview.md:7]: Okay I updated the language, please let me know if it sounds good.

File: tests/template_test.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
+# fmt: off
```
Comment (cbeauchesne) [inline tests/template_test.py:1]: I would recommend to entirely move this file away from the `tests` folder: 

* no chance to mess up with real tests
* you can easily put whatever you want inside it

Comment (colin-higgins) [reply tests/template_test.py]: Done

**Review by cbeauchesne — APPROVED** [review]

**Pull request merged**